### ISSUE TITLE
Grafana Logs "database is locked"

### ISSUE THREAD

**Issue by qq3884** [grafana/grafana#16638]
**What happened**:
At some point, Grafana gets an error.
Log out with the error "database is locked."
The following are some of Grafana's logs:
`lvl=eror msg="failed to look up user based on cookie" logger=context error="database is locked"`

**Environment**:
- Grafana version: grafana-6.1.4-1.x86_64
- Data source type & version: postgresql 10.3, graphite 9.x
- OS Grafana is installed on: Centos 6.9 amd64
- User OS & Browser: ANY
- Grafana plugins: piechart
- Others:

**Comment by torkelo** [comment]
Can you describe your setup more, options, number of users, alerts, dashboards, provisioned dashboards, alert evaluation frequency etc.

**Comment by seet61** [comment]
I have the same issue too. 
`t=2019-04-24T09:37:46+0300 lvl=eror msg="Could not load alerts" logger=alerting.ruleReader error="database is locked"
t=2019-04-24T09:37:46+0300 lvl=dbug msg="Scheduling update" logger=alerting.scheduler ruleCount=0
t=2019-04-24T09:37:55+0300 lvl=eror msg="Failed to get system stats" logger=metrics error="database is locked"
t=2019-04-24T09:37:56+0300 lvl=eror msg="Could not load alerts" logger=alerting.ruleReader error="database is locked"
t=2019-04-24T09:37:56+0300 lvl=dbug msg="Scheduling update" logger=alerting.scheduler ruleCount=0
t=2019-04-24T09:37:59+0300 lvl=dbug msg="auth token rotated" logger=auth affected=1 auth_token_id=46 userId=1
t=2019-04-24T09:37:59+0300 lvl=dbug msg="Updating last user_seen_at" logger=context userId=1 orgId=1 uname=admin user_id=1
`
Graphite, Grafana, Postgresql 9.6.

**Comment by seet61** [comment]
Another services working correctly with the same db.

**Comment by cuxcrider** [comment]
I am having a similar issue using influxdb.  Grafana works well for a while then all of sudden crashes and logs out.  Grafana v6.1.6 (commit: cf9cb45), Ubuntu 16.04.  

t=2019-05-14T00:36:04-0600 lvl=eror msg="Failed to get system stats" logger=metrics error="database is locked"
t=2019-05-14T00:36:05-0600 lvl=eror msg="Could not load alerts" logger=alerting.ruleReader error="database is locked"

**Comment by aggieben** [comment]
Same issue here.  Running on an EC2 instance: prometheus, grafana, and a couple of my own service.   Browser open to grafana, and that's it.  No other services hitting grafana.  Everything is fine for a few minutes, then at some point grafana craps itself with "database is locked" errors, and then it doesn't work, _and_ it won't start back up.

**Comment by cuxcrider** [comment]
When it says "database" is locked, is it referring to the influxdb data source or to the grafana database which I believe contains information about the dashboard, settings, etc and is an SQLite database by default?  I am running influxdb on a KVM VM and grafana on a separate KVM VM.  It would be nice to know if I should be trying to tune influx or grafana.

**Comment by aggieben** [comment]
@cuxcrider This is a problem with grafana's default sqlite db.

**Comment by cuxcrider** [comment]
Any idea if switching to postgres or mysql solves the problem?

**Comment by aggieben** [comment]
No, and I don't really have any interest in that, because part of what I'm working on is zero-config (or near-zero-config) setup for dev environments.  Having to configure a different DB as a backend is a non-starter for me.

**Comment by cuxcrider** [comment]
I hear ya there.  Would much prefer it to just "work".  I might give it a shot, I found some info here:
https://community.hortonworks.com/articles/33401/how-to-set-up-grafana-to-use-mysql-database-rather.html

**Comment by cuxcrider** [comment]
fyi I just set up grafana on a new VM and setup mysql using the guide I pasted above.  I then exported my dashboards as JSON and inserted them into my new grafana and so far, so good.  No crashing.  You can maybe convert your SQLite to mysql if you have a ton of stuff already setup, but my grafana was pretty minimal so I just exported my dashboards as json and then manually reconnected to data sources.

**Comment by truealex81** [comment]
I have similar issue when garafana suddenly stop working, and restarting doesn't fix this problem :(

t=2019-05-17T05:40:30+0000 lvl=eror msg="Could not load alerts" logger=alerting.ruleReader error="database is locked"
t=2019-05-17T05:40:40+0000 lvl=eror msg="Could not load alerts" logger=alerting.ruleReader error="database is locked"
t=2019-05-17T05:40:50+0000 lvl=eror msg="Could not load alerts" logger=alerting.ruleReader error="database is locked"
t=2019-05-17T05:40:59+0000 lvl=eror msg="failed to run garbage collect" logger=remotecache.database error="database is locked"
t=2019-05-17T05:40:59+0000 lvl=eror msg="Failed to get system stats" logger=metrics error="database is locked"
t=2019-05-17T05:41:00+0000 lvl=eror msg="Could not load alerts" logger=alerting.ruleReader error="database is locked"
t=2019-05-17T05:41:10+0000 lvl=eror msg="Could not load alerts" logger=alerting.ruleReader error="database is locked"
t=2019-05-17T05:41:20+0000 lvl=eror msg="Could not load alerts" logger=alerting.ruleReader error="database is locked"
t=20

**Comment by gsrcobr** [comment]
Same problem here. 
Grafana 6.1.6
CentOS 6.9

t=2019-05-17T11:41:00-0300 lvl=eror msg="failed to look up user based on cookie" logger=context error="database is locked"

Starts to happen when upgraded to Grafana 6.

**Comment by psby** [comment]
I've got same issue. 
Grafana send me 2 alerts tonight with Error message "Could not find datasource database is locked". 

Grafana 6.1.6 @ Debian 4.9.168 in docker (image grafana/grafana:6.1.6) 
InfluxDB

**Comment by kremers** [comment]
Same issue here v.6.1.6

**Comment by DanCech** [comment]
This seems to be an issue with sqlite itself, one way to recover the sqlite databse is to follow the procedure here: https://community.grafana.com/t/database-is-locked-unable-to-use-grafana-anymore/16557/2

The “database is locked” error indicates a problem with your sqlite database. This can happen if the database is left in an inconsistent state after a crash, or if there are problems with the disk.

One thing you can try is to dump the data from your existing db file into a new one, then swap it out. From inside your Grafana data directory (after shutting down Grafana):

```
sqlite3 grafana.db '.clone grafana-new.db'
mv grafana.db grafana-old.db
mv grafana-new.db grafana.db
```

**Comment by roidelapluie** [comment]
We have that a lot sinci grafana 6.2 (we use auth proxy)

**Comment by marefr** [comment]
@roidelapluie see #17247 for auth proxy specific problems with database is locked

**Comment by roidelapluie** [comment]
Ok we will upgrade to master ;) sorry for the noise, my github skills failed me this time!

**Comment by aggieben** [comment]
@DanCech isn't there anything Grafana can do to avoid or minimize this issue?  It seems pretty unusable to me from this vantage point.

**Comment by magicrobotmonkey** [comment]
Can confirm that cloning the sqlite fixed intermittent db locking logging out users on 6.2.1

**Comment by magicrobotmonkey** [comment]
6.2.1 (9e40b07) with sqlite backend, can confirm that cloning the db does not appear to help. I'm able to login and save some work, but intermittently get logged out with 
`t=2019-06-04T15:42:33+0000 lvl=eror msg="failed to look up user based on cookie" logger=context error="database is locked"`. 

I also see that my health monitor happened to scrape at the right timing and hit a 503:
`curl: (22) The requested URL returned error: 503 Service Unavailable`

**Comment by juanvmedrano** [comment]
This is a problem that affects many ...

Grafana after these errors is closing the session alone.

This is a very bad mistake because I can not put the monitors and have to start the session every hour and a half.

**Comment by cuxcrider** [comment]
I agree, sqlite was simply not a usable solution.  Grafana would log me out
after only a few minutes of use.  It would remain running, but I would have
to log in again.

for what it is worth I switched to mysql for grafana over two weeks ago and
have not had a single error even with multiple users and multiple
datasources.

**Comment by davewat** [comment]
The Grafana default for SQLite journaling is the Rollback Journal (DELETE mode by PRAGMA), which originates from go-sqlite3.  In addition, go-sqlite3 forces it back to DELETE mode even if you manually set to WAL (take DB offline and change PRAGMA journal_mode=wal;)

Switching to WAL (Write Ahead Log) mode reduces many opportunities for lock contention on the DB, and would likely provide a performance increase(https://www.sqlite.org/wal.html).  We are doing this with a fork to relieve lock errors that appeared with heavy Grafana Alert queries.  Might be worth looking into moving to WAL mode.

**Comment by DanCech** [comment]
@davewat You can pass the connection string directly, so something like this should work:
```
[database]
type=sqlite3
connection_string=file:data/grafana.db?cache=private&mode=rwc&_journal_mode=WAL
```

**Comment by davewat** [comment]
> @davewat You can pass the connection string directly, so something like this should work:
> 
> ```
> [database]
> type=sqlite3
> connection_string=file:data/grafana.db?cache=private&mode=rwc&_journal_mode=WAL
> ```

That is excellent - thanks!  That is much better than running a custom build just for WAL mode!  I noticed you also used private cache in your example, but source shows shared cache (we also are running private cache).

**Comment by DanCech** [comment]
In 6.1.6 it was made configurable and the default changed to private, but if you use the `connection_string` setting that is just passed directly to `xorm` and through to `go-sqlite3` and anything else in the `[database]` block (besides `type`) is pretty much ignored.

**Comment by magicrobotmonkey** [comment]
yea switching to WAL journal made a big difference for me. Not only is it not logging me out, its much, much faster.

**Comment by juanvmedrano** [comment]
> @davewat You can pass the connection string directly, so something like this should work:
> 
> ```
> [database]
> type=sqlite3
> connection_string=file:data/grafana.db?cache=private&mode=rwc&_journal_mode=WAL
> ```


Where can I apply this configuration?

I want to try this.

**Comment by DanCech** [comment]
@juanvmedrano you'd include it in your grafana configuration file

**Comment by EdisonRobertoJr** [comment]
> @davewat You can pass the connection string directly, so something like this should work:
> 
> ```
> [database]
> type=sqlite3
> connection_string=file:data/grafana.db?cache=private&mode=rwc&_journal_mode=WAL
> ```

Nice, worked to me, but you have to change the "connection_string=file:data/grafana.db" to your grafana.db path.

**Comment by Georege** [comment]
Same issue here v.6.1.6

**Comment by zegerius** [comment]
Same issue here on: version=6.2.5 commit=6082d19 branch=HEAD compiled=2019-06-25T17:56:19+0000

**Comment by sipicsg** [comment]
Same with docker image grafana/grafana:6.2.5 .

**Comment by brunetton** [comment]
Same here (sorry for the noise)
Grafana v6.1.1 (eff01d2)
I'm totaly new to Grafana, just added a new empty dashboard + one panel. This errors stops me from using grafana. 

I tried [Switching to WAL (Write Ahead Log) as suggested by @davewat](https://github.com/grafana/grafana/issues/16638#issuecomment-498811755) and so far so good :) (using it for 2 hours now, created dashboards, panels, no problem !)

**Comment by Quicksy** [comment]
Same here, with grafana in kubernetes with  image: 'grafana/grafana:6.2.2'

**Comment by AvinKumar** [comment]
Facing the same "database is locked" issue on my Grafana v6.4.2 setup installed in SUSE 12SP3. 
Noticed this issue only after I upgraded the Grafana version from v6.0.x to v6.4.2

-Avin

**Comment by mbolek** [comment]
grafana 6.4.3 on amazon ami 2
fresh install, using default sqlite
```t=2019-10-29T09:26:26+0000 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=302 remote_addr=XXXX time_ms=0 size=29 referer=
t=2019-10-29T09:26:38+0000 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=302 remote_addr=XXXX time_ms=0 size=29 referer=
t=2019-10-29T09:26:43+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2019-10-29T09:26:56+0000 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=302 remote_addr=XXXX time_ms=0 size=29 referer=
t=2019-10-29T09:27:08+0000 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=302 remote_addr=XXXX time_ms=0 size=29 referer=
t=2019-10-29T09:27:17+0000 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=302 remote_addr=XXXX 1 time_ms=0 size=29 referer=
t=2019-10-29T09:27:23+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2019-10-29T09:27:26+0000 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=302 remote_addr=XXXX time_ms=0 size=29 referer=
t=2019-10-29T09:27:38+0000 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=302 remote_addr=XXXX time_ms=0 size=29 referer=
t=2019-10-29T09:27:43+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2019-10-29T09:27:56+0000 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=302 remote_addr=XXXX time_ms=0 size=29 referer=
t=2019-10-29T09:28:03+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2019-10-29T09:28:03+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2019-10-29T09:28:08+0000 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=302 remote_addr=XXXX time_ms=0 size=29 referer=
t=2019-10-29T09:28:10+0000 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=302 remote_addr=XXXX time_ms=0 size=29 referer=
```
It's rather annoying, since it'll often fail on listing all dashboards, displaying only a portion of the list, on the other occasion it'll display the list but then will fail on loading a dashboard with 404, upon page reload it's fine.
I have used the tip from https://github.com/grafana/grafana/issues/16638#issuecomment-498818278
but doesn't seem to change much in my case.
As for the setup:
~30 dashboards
~ 5 users (not simultaneous)
no alerts
I'll try with postgres but it has already been mentioned to also have issues

**Comment by mbolek** [comment]
@marefr hey, I did and still had issues. Was able to work around this problem by removing the DB (so no migration etc) and start fresh

**Comment by tgranqvist** [comment]
I'm seeing this same error when deploying to Azure Container Instances and mounting a volume from Azure File Share (see issue linked by marefr). I've tried the solution from @DanCech [here]. This seem to have worked for several people in this thread. Unfortunately not working for me. I see in the container log that the default settings are overridden and the database file is created with the name from my ini file, but still get that same error: _"Service init failed: Migration failed err: database is locked"_.

My ini file in its entirety:
```
[database]
type=sqlite3
connection_string=file:/var/lib/grafana/mygrafana.db?_journal_mode=WAL
```

Any pointers on what to do? Azure Files volumes are SMB3, does that have anything to do with the issue?

My yaml deployment file: [deploy-gvt-11.yaml.txt](https://github.com/grafana/grafana/files/3876587/deploy-gvt-11.yaml.txt)
Deployment log file: [aci-grafana.log](https://github.com/grafana/grafana/files/3876595/aci-grafana.log)

Any help is much appreciated!

[here]: https://github.com/grafana/grafana/issues/16638#issuecomment-498818278

**Comment by Iziman95** [comment]
Hi everyone,

I have an issue where my dashboard provisioning is not working properly. 
It load some of my .json files and doesn't load some other, which randomly cause a "Dashboard not found" error message. I have around 25 .json files.

In my investigation of this issue I have noticed that the grafana.log is full of "Database locked" log message. This message occur around 10 times every 10 seconds.
I am also noticing that in the grafana.db the dashboard_provisioning table is always updated but randomly missing some .json files.
I think the database lock is related to this issue. Am I wrong ? Is there a way to figure what cause the db lock ?

Environment:

Grafana version: Grafana v6.5.1
Grafana DB :  sqlite
OS Grafana is installed on: RHEL 7.4
Users : less than 10

Any help is much appreciated

**Comment by marefr** [comment]
@Iziman95 could be related to using same uid in multiple dashboards?

**Comment by Iziman95** [comment]
> @Iziman95 could be related to using same uid in multiple dashboards?

thanks but I have already checked this when I was searching for an answer here, all my uids are unique

**Comment by Iziman95** [comment]
@marefr The grafana documentation describes how provisioning is supposed to work :
"Dashboards will be reloaded when the json files changes".
But it is like they are permanently reloaded every 5-10 seconds while there is no change on these files. I can confirm this by watching the table "dashboard_provisioning" and seeing that the "id" increases indefinitely each time I perform a select (current value is 2005572) for the same rows.
Is there a way to take .json files only at the start of grafana server and not after ?

**Comment by marefr** [comment]
@Iziman95 you must have configured something wrong. Does any of your dashboard have an `id` with a numeric value? Are you by any chance multiple paths configured where one of them is a sub directory of the other? If that doesn't help I would suggest open a new issue with provided provisioning config and dashboards so that we can look at it separately from this issue..

**Comment by Iziman95** [comment]
@marefr ok so I have done many tests with fresh dashboards in different paths and checked all my configuration and I have found that the root cause was the property providers name which was the same in all of my yaml file. 
After editing my yaml files to have an unique name my issue is solved and I have no more "database locked" in my logs. My database size is also huge now because the table dashboard_version was permanently updated with my +20 dashboard every 10 seconds.

Thank you for your help

**Comment by pirminjanka** [comment]
i had the same problem with grafana 6.5.1 on kubernetes. I did the following inside the container:

`
mv /var/lib/grafana/grafana.db /var/lib/grafana/grafana.db.old
rm /var/lib/grafana/grafana.db
mv /var/lib/grafana/grafana.db.old /var/lib/grafana/grafana.db
`

after this steps it worked fine.
Maybe my issue was because the grafana container switched to a different kubernetes node and therefore the database was locked

**Comment by dylandragon93** [comment]
If you're trying to get Grafana deployed via an Azure App Service container, [this](https://github.com/grafana/grafana/issues/20549#issuecomment-584857241) might be of help.

This issue has also been brought up to Microsoft in https://github.com/MicrosoftDocs/azure-docs/issues/47130.

**Comment by lsoica** [comment]
I'm getting the same error logs during IO heavy workloads (I can reproduce it by running a dd on the same disk as grafana's). My disk is a remote disk (Ceph RBD) and during a dd I get 100% util and a very high latency (await > 10s). 

`
lvl=eror msg="Failed to get system stats" logger=metrics error="database is locked"
lvl=eror msg="failed to search for dashboards" logger=provisioning.dashboard type=file name=default error="database is locked"
lvl=eror msg="Could not load alerts" logger=alerting.ruleReader error="database is locked"
`

**Comment by lsoica** [comment]
> I'm getting the same error logs during IO heavy workloads (I can reproduce it by running a dd on the same disk as grafana's). My disk is a remote disk (Ceph RBD) and during a dd I get 100% util and a very high latency (await > 10s).
> 
> `lvl=eror msg="Failed to get system stats" logger=metrics error="database is locked" lvl=eror msg="failed to search for dashboards" logger=provisioning.dashboard type=file name=default error="database is locked" lvl=eror msg="Could not load alerts" logger=alerting.ruleReader error="database is locked"`

_journal_mode=WAL didn't work for me. I had to add _busy_timeout in the connection string aswell.

**Comment by kuldeeps88** [comment]
Environment:
Grafana version: Grafana v6.3.5
Grafana DB : sqlite

With Grafana as kubernetes deployment, Without adding any data source or dashboards, Grafana Docker container itself is throwing msg="Database locked, sleeping then retrying" logger=sqlstore error=" database is locked" retry=0

This issue we observed after upgrading from 4.6.3 to 6.3.5.

**Note: Grafana Docker container itself is throwing error without any data source or dashboard being added.

Also, Grafana is deployed as a Kubernetes deployment so it's getting scheduled on numerous worker nodes with the same error.

`t=2020-03-09T19:03:42+0000 lvl=info msg="HTTP Server Listen" logger=http.server address=0.0.0.0:3000 protocol=http subUrl=/grafana socket=
t=2020-03-09T19:03:42+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:42+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=1
t=2020-03-09T19:03:43+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:43+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:43+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:43+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:43+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:43+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=1
t=2020-03-09T19:03:43+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:43+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:43+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=1
t=2020-03-09T19:03:44+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:44+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:44+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:44+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:44+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:44+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:44+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-03-09T19:03:44+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
`

**Comment by f-ld** [comment]
Got that issue on my kubernetes deployment with grafana 6.0.1. It was almost systematic, with some alerts not loading (well appearing in dashboard but not in list of alerts and remaining as not active in dashboard). Mote: we are using provisioned dashboards, alerts, etc ...

Changed to latest version, grafana 6.7.1, and it look way better, issue not reproduced so far. I will update this comment if I see it again in the next future. But I would recommend all having this issue to update grafana

**Comment by andloh** [comment]
Same issue on the latest grafana docker image. Using Azure files with mountoption uid/gid 65534
`msg="Server shutdown" logger=server reason="Service init failed: Migration failed err: database is locked"`

Tried `connection_string=file:/var/lib/grafana/grafana.db?cache=private&mode=rwc&_journal_mode=WAL`, not working.

**Comment by beingyash** [comment]
@torkelo Any update?

**Comment by f-ld** [comment]
Almost 1 month and a half later after https://github.com/grafana/grafana/issues/16638#issuecomment-606605113, with grafana 6.7.1 on k8s cluster created with kops on AWS, still not reproduced even once.

**Comment by beingyash** [comment]
I have v6.7.3 and am still facing this issue.

**Comment by sukhan85** [comment]
is this issue resolved? We are facing the similar issue.

**Comment by KursLabIgor** [comment]
same issue EC2 + EFS mounted.

**Comment by sorriso93** [comment]
same problem here

**Comment by KursLabIgor** [comment]
i solved it temporary moved data folder to EBS storage, and now planning move to MySQL DB ad store it in EFS

**Comment by sorriso93** [comment]
> i solved it temporary moved data folder to EBS storage, and now planning move to MySQL DB ad store it in EFS

what is EBS storage? I have the sqllite db of grafana on external hard disk with the raspberry, disabled login page and enabled anonymous user. no result, same problem!

**Comment by pimjansen** [comment]
Same here on Azurefile

**Comment by cwrau** [comment]
It's happening to us as well, but we're not using persistent storage, just an `emptyDir`

On AKS

**Comment by ilietrasca** [comment]
I encounter the same problem with version 7.0.3.

```
t=2020-06-19T19:55:33+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-06-19T22:24:15+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-06-20T01:23:06+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-06-20T04:12:02+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
t=2020-06-20T04:12:02+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=1
t=2020-06-20T07:42:16+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=0
```

The error received in the browser:
`A cookie associated with a resource at http://monitoring.sre.idemia.io/ was set with `SameSite=None` but without `Secure`. A future release of Chrome will only deliver cookies marked `SameSite=None` if they are also marked `Secure`. You can review cookies in developer tools under Application>Storage>Cookies and see more details at https://www.chromestatus.com/feature/5633521622188032.`

Can you please provide an workaround for this issue ?

**Comment by DSalvigni** [comment]
Hello Guys,
i am running the 7.0.3 as docker container directly from docker HUB in a Azure App Service but since some days it stopped to work. I have the database saved in Azure File Share but any time I tired to logon i receive the error of "Database Locked". More in details:

2020-06-24T11:20:41.558497227Z t=2020-06-24T11:20:41+0000 lvl=info msg="Connecting to DB" logger=sqlstore dbtype=sqlite3
2020-06-24T11:20:41.559484232Z t=2020-06-24T11:20:41+0000 lvl=info msg="Starting DB migration" logger=migrator
2020-06-24T11:20:46.689980794Z t=2020-06-24T11:20:46+0000 lvl=eror msg="Server shutdown" logger=server reason="Service init failed: Migration failed err: database is locked"

Any workaround??

**Comment by tynor88** [comment]
@DSalvigni we are also experiencing the same issues the past couple of days... No workaround yet.

**Comment by DSalvigni** [comment]
So I fixed in this way, but before start please backup locally the sqlite3 database (grafana.DB).

1) GF_DATABASE_TYPE=postgresql (or you can do the same in MySQL)
2) Created a DB called grafana in a postgresql Server
3) Set up the following paramenters for the container (in App Service):
GF_DATABASE_HOST=
GF_DATABASE_NAME=
GF_DATABASE_PASSWORD=
GF_DATABASE_TYPE=
GF_DATABASE_USER=
3) Moved the mounted persistent storage "/var/lib/grafana" under Azure File Storage (and I got rid of the Blob COntainer).
5) Deployed from Docker HUB the new version of Grafana

Now you have a new deployed version of grafana clena up an running. You need to port the data from the old DB to the new one and for this I simply dumped the content of the tables of grafana,DB into the new created grafana DB in postgresql.

To access grafana.DB i used "SQLiteDatabaseBrowserPortable" and to move the data this "https://www.dbsofts.com/". The trial version limits the dumb to 10.000 lines but more than enough to move the following table:

alert*
org*
dashboard* (not necessary if you have dumped already your dashboard as JSON)
data_source (the most important)

Let me know if you need some more questions.

**Comment by kaisert** [comment]
Experiencing the same issue. Grafana randomly logs out and only shows its error page. In the logs it says
```
2020-07-07T02:04:03.442792383Z t=2020-07-07T02:04:03+0000 lvl=info msg="Database locked, sleeping then retrying" logger=sqlstore error="database is locked" retry=8166
```
Restarting the container brings back Grafana for a while.

Using Grafana 7.0.5 docker image deployed as an app service on azure. /var/lib/grafana is connected to a file share. 
`GF_DATABASE_URL` is set to `sqlite3:///var/lib/grafana/grafana.db?cache=private&mode=rwc&_journal_mode=WAL`

**Comment by DSalvigni** [comment]
> `GF_DATABASE_URL` is set to `sqlite3:///var/lib/grafana/grafana.db?cache=private&mode=rwc&_journal_mode=WAL`

This did not worked properly for me: DB remain always locked, so I moved directly to an external DB and the problem is solved, plus finally i can connect BI to link SLA to Events backlog 👍

**Comment by sacrelege** [comment]
Displaying a dashboard with 10s refresh (graphite datasource). Logouts happen random, between 10min and 2-3 hours.

Fresh installation (.deb) of Grafana v7.3.0-test (98b94d3824) on a fresh Ubuntu 20.04.1 LTS. Grafana is pretty much on default settings (sqlite).

hardware: 12core 3900x, 64GB ram, HP EX950 nvme (2.9GB/s read/write speed).

log entries: msg="Failed to look up user based on cookie" logger=context error="user token not found"

**Comment by biuboombiuboom** [comment]
I guess this problem is caused by the slow backend query service。
![Uploading 企业微信截图_16037649943587.png…]()

**Comment by biuboombiuboom** [comment]
> I guess this problem is caused by the slow backend query service。
> ![企业微信截图_16037649943587](https://user-images.githubusercontent.com/7258606/97248886-b5607000-183d-11eb-90ef-70a5af3a3a25.png)
> 
> Or maybe rotated twice when T1

log

t=2020-10-27T04:07:23+0000 lvl=info msg="rotate new token sucess" logger=auth newToken=cb4d13f08b6f4e49c787098a197c56c9 oldToken=2a12712084517785a7dffc77142bf67a clientIp=10.181.34.189
t=2020-10-27T04:17:24+0000 lvl=info msg="rotate new token sucess" logger=auth newToken=f7eb1aeeb6d76daa5b0bf06aa56d3990 oldToken=cb4d13f08b6f4e49c787098a197c56c9 clientIp=10.181.34.189

**Comment by biuboombiuboom** [comment]
Use channel rotate token control concurrency。
OR
Add lock 

type UserToken struct {
     sync.Mutex
}

auth_token.go
func (s *UserAuthTokenService) TryRotateToken(ctx context.Context, token *models.UserToken,
	clientAddr, userAgent string) (bool, error) {
	token.Lock()
	defer token.Unlock()
}

**Comment by pimjansen** [comment]
In Azure i finally got it working by running this command on the database:

```
sqlite3 grafana.db 'pragma journal_mode=wal;'
```

After that i overwritten the database using the Azure Storage explorer and all is running fine from there

**Comment by lobocobra** [comment]
same issue here with grafana and influxdb

**Comment by KursLabIgor** [comment]
> > i solved it temporary moved data folder to EBS storage, and now planning move to MySQL DB ad store it in EFS
> 
> what is EBS storage? I have the sqllite db of grafana on external hard disk with the raspberry, disabled login page and enabled anonymous user. no result, same problem!

This is hard disk storage for VM (EC2) in AWS it is called like EBS

**Comment by petersohn** [comment]
There seems to be multiple issues here, actually. One is the database locking issue that Grafana sometimes fails to access its user database. There is an error handling issue too. If this error happens, the server returns with a 401 status code, for which the client does a logout. However, this should not be a 401 status code, but instead of an 500 or something similar, because the problem is not with the client's authentication, but with the server database. Then the client would occasionally fail a query, but it would not log out.

**Comment by cameleon77** [comment]
+1

same issue here, just upgraded to Grafana 7.3.4 from 5.x.x and started getting:
`t=2020-12-05T14:00:00+0100 lvl=eror msg="Failed to look up user based on cookie" logger=context error="user token not found"`
some time after logged in

Grafana default settings with sqlite3
installed on some old Ubuntu 16.04

**Comment by fuomag9** [comment]
Same issue here, rpi4 64bit grafana 7.3.6


```
  |   | Jan 15 12:27:54 raspi4 grafana-server[2259]: t=2021-01-15T12:27:53+0100 lvl=eror msg="Failed to get system stats" logger=metrics error="database is locked"
  |   | Jan 15 12:27:54 raspi4 grafana-server[2259]: t=2021-01-15T12:27:53+0100 lvl=eror msg="handler.notifier.SendIfNeeded failed" logger=alerting.resultHandler err="database is locked"
  |   | Jan 15 12:27:54 raspi4 grafana-server[2259]: t=2021-01-15T12:27:53+0100 lvl=eror msg="Failed to get alert notifiers" logger=alerting.notifier error="database is locked"
  |   | Jan 15 12:27:54 raspi4 grafana-server[2259]: t=2021-01-15T12:27:53+0100 lvl=eror msg="handler.notifier.SendIfNeeded failed" logger=alerting.resultHandler err="database is locked"
  |   | Jan 15 12:27:54 raspi4 grafana-server[2259]: t=2021-01-15T12:27:53+0100 lvl=eror msg="handler.notifier.SendIfNeeded failed" logger=alerting.resultHandler err="database is locked"
  |   | Jan 15 12:27:54 raspi4 grafana-server[2259]: t=2021-01-15T12:27:53+0100 lvl=eror msg="Failed to get alert notifiers" logger=alerting.notifier error="database is locked"
  |   | Jan 15 12:27:53 raspi4 grafana-server[2259]: t=2021-01-15T12:27:53+0100 lvl=eror msg="Failed to get alert notifiers" logger=alerting.notifier error="database is locked"


```

**Comment by AnandDevan** [comment]
+1 

Same issue in our setup that uses sqlite

It looks like this new setting is forcing an auth refresh every 10m by default [token_rotation_interval_minutes](https://grafana.com/docs/grafana/v7.3/administration/configuration/#token_rotation_interval_minutes) setting 

Is this the reason for users getting logged off with the error below? 
 ```
t=2020-12-05T14:00:00+0100 lvl=eror msg="Failed to look up user based on cookie" logger=context error="user token not found"
```

**Comment by dokuboyejo** [comment]
Had to downgrade from v7.4.3 to v6.3.7 to resolve constant DB lock
As some others mentioned, seems issues has been recorded using SQLite / Postgres / MySQL
@AnandDevan wrt token refresh...the lock behavior sometimes occurs many times in space of a minute, suggesting it might not necessarily be that

**Comment by AnandDevan** [comment]
I want to post some resources for the folks that end up here.

To migrate from sqlite to postgres we used the following:
https://wbhegedus.me/migrating-grafanas-database-from-sqlite-to-postgres/
https://github.com/haron/grafana-migrator

**Comment by gcurti** [comment]
> In Azure i finally got it working by running this command on the database:
> 
> ```
> sqlite3 grafana.db 'pragma journal_mode=wal;'
> ```
> 
> After that i overwritten the database using the Azure Storage explorer and all is running fine from there

I had the same issue but in baremetal environment and worked it great, thanks!

**Comment by kisshore** [comment]
Did anyone tried  v7.5.5+ & able to replicate issue?
I see there is commit for v7.5.5 https://github.com/grafana/grafana/pull/32999

Also this suggestion looks good https://github.com/grafana/grafana/issues/16638#issuecomment-707372085

**Comment by tgritter** [comment]
Receiving this `database is locked` issue using influxDB 1.8.9 and Grafana 8.1.3. Any thoughts on how to solve this issue for influxDB?

**Comment by cloudcafetech** [comment]
Seems ```sqlite3 grafana.db 'pragma journal_mode=wal;'``` works for others.

Can anyone tell me how to fix in kubernetes statefull deployment ?

**Comment by lafrech** [comment]
Same issue here with Grafana 8.3.3.

I've been trying to apply the suggestion found in this thread (and in #17912) about using WAL journal mode.

```
[database]
type=sqlite3
connection_string=file:data/grafana.db?cache=private&mode=rwc&_journal_mode=WAL
```

Unfortunately, although the config is standard (official Debian packages), with `data` being `/var/lib/grafana` and the database file being `grafana.db`, the database path doesn't seem to be recognized.

The logs stop at

    msg="Creating SQLite database file" logger=sqlstore path=grafana.db

I also tried by passing the absolute path instead or "data", to no avail.

Also, I couldn't find any documentation about `connection_string`.

I would suggest to reopen #17912.

**Comment by pimjansen** [comment]
> Same issue here with Grafana 8.3.3.
> 
> 
> 
> I've been trying to apply the suggestion found in this thread (and in #17912) about using WAL journal mode.
> 
> 
> 
> ```
> 
> [database]
> 
> type=sqlite3
> 
> connection_string=file:data/grafana.db?cache=private&mode=rwc&_journal_mode=WAL
> 
> ```
> 
> 
> 
> Unfortunately, although the config is standard (official Debian packages), with `data` being `/var/lib/grafana` and the database file being `grafana.db`, the database path doesn't seem to be recognized.
> 
> 
> 
> The logs stop at
> 
> 
> 
>     msg="Creating SQLite database file" logger=sqlstore path=grafana.db
> 
> 
> 
> I also tried by passing the absolute path instead or "data", to no avail.
> 
> 
> 
> Also, I couldn't find any documentation about `connection_string`.
> 
> 
> 
> I would suggest to reopen #17912.
> 
> 
> 
> 

Did you convert the db after creation? Since i did this all works like a charm

**Comment by lafrech** [comment]
> Did you convert the db after creation?

I didn't read about conversion. I expected it to work right away even on an existing DB.

Any pointer to the missing step? What would be the conversion command?

**Comment by pimjansen** [comment]
> > Did you convert the db after creation?
> 
> 
> 
> I didn't read about conversion. I expected it to work right away even on an existing DB.
> 
> 
> 
> Any pointer to the missing step? What would be the conversion command?

I used below command on the database

sqlite3 grafana.db 'pragma journal_mode=wal;'

After this all works as expected

**Comment by lafrech** [comment]
I saw this command but didn't get the point.

So the process would be

```
stop grafana
backup db
sqlite3 grafana.db 'pragma journal_mode=wal;'
edit grafana.ini to add the connection_string line
restart grafana
```

I just tried that but got the same error. The error message about creating a database makes me think the connection_string is not interpreted correctly and grafana doesn't find my grafana.db.

This could be specific to my setup and I wouldn't like to send more noise to this issue. Yet I'm kinda stuck since I can't find much more about connection_string.

Hence my proposal to go on with #17912. It is old but the situation doesn't seem to have changed since then and the fix is not straightforward. At least not enough for me.

**Comment by JedenFalls** [comment]
The same questions, 
Ubuntu 18.04(Azure VM)
1. I stopped grafana 
2. implemented command ` sqlite3 grafana.db 'pragma journal_mode=wal;'`
3. start grafana, but problem with alerts db is locked error still there 

I've tried add connection string 
```
[database]
type=sqlite3
connection_string=file:data/grafana.db?cache=private&mode=rwc&_journal_mode=WAL
```
but in this case grafana can't start

```
● grafana-server.service - Grafana instance
   Loaded: loaded (/usr/lib/systemd/system/grafana-server.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Fri 2022-01-14 06:46:23 UTC; 3s ago
     Docs: http://docs.grafana.org
  Process: 10933 ExecStart=/usr/sbin/grafana-server --config=${CONF_FILE} --pidfile=${PID_FILE_DIR}/grafana-server.pid --packaging=deb
 Main PID: 10933 (code=exited, status=1/FAILURE)

Jan 14 06:46:23 vmgrafana systemd[1]: grafana-server.service: Main process exited, code=exited, status=1/FAILURE
Jan 14 06:46:23 vmgrafana systemd[1]: grafana-server.service: Failed with result 'exit-code'.
Jan 14 06:46:23 vmgrafana systemd[1]: grafana-server.service: Service hold-off time over, scheduling restart.
Jan 14 06:46:23 vmgrafana systemd[1]: grafana-server.service: Scheduled restart job, restart counter is at 5.
Jan 14 06:46:23 vmgrafana systemd[1]: Stopped Grafana instance.
Jan 14 06:46:23 vmgrafana systemd[1]: grafana-server.service: Start request repeated too quickly.
Jan 14 06:46:23 vmgrafana systemd[1]: grafana-server.service: Failed with result 'exit-code'.
Jan 14 06:46:23 vmgrafana systemd[1]: Failed to start Grafana instance.
```

**Comment by JedenFalls** [comment]
Fighting the same issue, in my case grafana.db located in /var/lib/grafana.db
therefore need run sqlite command:
`sudo sqlite3 /var/lib/grafana/grafana.db 'pragma journal_mode=wal;'`

or you can do like this
`sudo sqlite3 /var/lib/grafana/grafana.db`

then check that you are in correct db place
```
sqlite> .show
        echo: off
         eqp: off
     explain: auto
     headers: off
        mode: list
   nullvalue: ""
      output: stdout
colseparator: "|"
rowseparator: "\n"
       stats: off
       width:
    filename: /var/lib/grafana/grafana.db
```

then set wal
```
sqlite> pragma journal_mode=wal;
sqlite>.quit
```

**Comment by Umby79** [comment]
I ran into the same issue while using app service in azure.

I solved the issue by changing the app planning parameter Capacity from 2 to 1.
That was causing the DB lock for me.

Hope this helps someone else too....

**Comment by DSalvigni** [comment]
> I ran into the same issue while using app service in azure.
> 
> I solved the issue by changing the app planning parameter Capacity from 2 to 1. That was causing the DB lock for me.
> 
> Hope this helps someone else too....

WHere exactly it is this setting in app service???

**Comment by Umby79** [comment]
> > I ran into the same issue while using app service in azure.
> > I solved the issue by changing the app planning parameter Capacity from 2 to 1. That was causing the DB lock for me.
> > Hope this helps someone else too....
> 
> WHere exactly it is this setting in app service???

Hi there, 

It is in the "app service plan"
In the json file that I use to deploy it is called capacity:

            "sku": {
                "name": "P1v2",
                "tier": "PremiumV2",
                "size": "P1v2",
                "family": "Pv2",
                "capacity": 1

In the portal you can find this parameter under the app service plan menu-->settings-->properties.
There it is called "instance count". Not sure why the name is so different. I am new to Azure.

Just to be clear, you should also first change the connection string to:
connection_string=file:data/grafana.db?cache=private&mode=rwc&_journal_mode=WAL
and create the DB using the WAL mode as suggested in this same discussion.

However, for me it started to work only after I also change the capacity parameter as I mentioned.

Good luck and let me know if it is working

**Comment by Dhyanesh97** [comment]
![image](https://user-images.githubusercontent.com/89400147/157373235-fe13d82e-2070-4ec4-ad2f-c0defbe19324.png)

We are also facing same issue. Error occurred for both cloud-watch (5 rules) and influx-db (25 rules) as data source.

Grafana 8.2.1
influxdb 2.0

**Comment by Lithimlin** [comment]
Same issue here.
* NixOS 21.11
* Grafana 8.4.4

```
Mar 25 07:35:16 <host> grafana-start[33992]: logger=sqlstore t=2022-03-25T07:35:16.78+0100 lvl=eror msg="Failed getting data source" err="database is locked" uid=PAA5C29273E02624E id=0 name= orgId=1
Mar 25 07:35:18 <host> grafana-start[33992]: logger=sqlstore t=2022-03-25T07:35:17.89+0100 lvl=eror msg="Failed getting data source" err="database is locked" uid=PAA5C29273E02624E id=0 name= orgId=1
Mar 25 07:35:19 <host> grafana-start[33992]: logger=sqlstore t=2022-03-25T07:35:19+0100 lvl=eror msg="Failed getting data source" err="database is locked" uid=PAA5C29273E02624E id=0 name= orgId=1
```
This also causes various alerts to fire with no values. I have not yet found a pattern in which alerts will be affected.

**Comment by daverstam** [comment]
> Indeed, I read [the docs](https://sqlite.org/wal.html) more thoroughly (which I should have done in the first place). They read:
> 
> > Unlike the other journaling modes, PRAGMA journal_mode=WAL is persistent. If a process sets WAL mode, then closes and reopens the database, the database will come back in WAL mode. In contrast, if a process sets (for example) PRAGMA journal_mode=TRUNCATE and then closes and reopens the database will come back up in the default rollback mode of DELETE rather than the previous TRUNCATE setting.
> > The persistence of WAL mode means that applications can be converted to using SQLite in WAL mode without making any changes to the application itself. One has merely to run "PRAGMA journal_mode=WAL;" on the database file(s) using the command-line shell or other utility, then restart the application.
> > The WAL journal mode will be set on all connections to the same database file if it is set on any one connection.
> 
> My conclusions:
> 
> * The `connection_string` method doesn't seem to work. Perhaps I did it wrong or it is obsolete, but from the logs it looks like Grafana doesn't find the existing DB file.
> * Setting WAL manually using sqlite3 should make the change permanent. Unless, I guess, Grafana explicitly specifies a different connection method on connection.
> 
> I just tried the sqlite3 method. I'll report here later on.
> 
> **Jan 26 edit**: I did that change on Jan 14th. The error seems to have disappeared since Jan 22nd. It's hard to conclude that it disappeared as a consequence of this change, as I can't explain the 8 days delay. I'll see if the problem reappears.

As I understand it setting wal in the db should be persistent and can't be changed by any application/connection.
"Unlike the other journaling modes, [PRAGMA journal_mode=WAL](https://sqlite.org/pragma.html#pragma_journal_mode) is persistent."

**Comment by lafrech** [comment]
Yes, I was unclear. I meant the change is permanent indeed and it should apply (by default) to all new Grafana connections, unless Grafana explicitly specifies a different connection method (overriding this new permanent default).

BTW the issue never reappeared here so the sqlite3 method did the trick. I just can't explain why the issue persisted a few days after the fix.

**Comment by daverstam** [comment]
I just did the wal change and still get this in the logs:
`logger=sqlstore t=2022-04-07T10:55:55.1+0200 lvl=info msg="Database locked, sleeping then retrying" error="database is locked" retry=0`

I recently started to investigate if Grafana ngalert could be a contender to Alertmanager since I feel that Grafana has a nice UI however I'm running in to weird problems. One issue particular where all my alerts would get resolved and then a couple of minutes later gets in alert/firing mode again. I have the same alerts configured in Alertmanager and they stay in alert/firing mode there which is the correct state.

I'm suspecting this is caused by database locks but not sure really

**Comment by daverstam** [comment]
Update: It's now been 4days since I changed the DB to use wal and it seems to have solved the issue I had where all my alerts would suddenly get in resolved state.

I can still see this in the logs however it does not seem to affect anything since I don't have any issues.
`logger=sqlstore t=2022-04-11T10:10:56.29+0200 lvl=info msg="Database locked, sleeping then retrying" error="database is locked" retry=0`

**Comment by oleksandrnezhynskyi** [comment]
For a Grafana that deployed into Kubernetes cluster is not possible to execute 'sqlite3' command.
In this case will help adding this into helm chart:
`grafana.ini:
    database:
      cache_mode: shared`

**Comment by kavaka123** [comment]
Any suggestions on how to fix this in k8s deployment?
I added the parameter "cache_mode: shared" like @k0ntrik said, and it worked for 5 days and started happening again. The error in logs is:
`logger=sqlstore t=2022-12-04T10:10:56.29+0200 lvl=info msg="Database locked, sleeping then retrying" error="database is locked" retry=0`

**Comment by strongpapazola** [comment]
![image](https://user-images.githubusercontent.com/48545591/208859966-0f00a9b4-2319-4b75-9934-ec9f3623bf7e.png)

**Comment by DLoT** [comment]
i have the same issue, happens on save but also on alert evaluation

logger=ngalert.scheduler rule_uid=aLHc4khVz org_id=1 version=3 attempt=0 now=2023-01-05T11:03:00Z t=2023-01-05T11:03:10.014606251Z level=error msg="Failed to build rule evaluator" error="failed to build query 'A': database is locked"

![image](https://user-images.githubusercontent.com/2110405/210765231-ee0c2b55-36ba-4553-b568-df4077a10d9c.png)

**Comment by jehutywong** [comment]
Version 9.3.6 is also faceing the same issue, when building alert rules with Prometheus datasource.

**Comment by ianmuge** [comment]
> I saw this command but didn't get the point.
> 
> So the process would be
> 
> ```
> stop grafana
> backup db
> sqlite3 grafana.db 'pragma journal_mode=wal;'
> edit grafana.ini to add the connection_string line
> restart grafana
> ```
> 
> I just tried that but got the same error. The error message about creating a database makes me think the connection_string is not interpreted correctly and grafana doesn't find my grafana.db.
> 
> This could be specific to my setup and I wouldn't like to send more noise to this issue. Yet I'm kinda stuck since I can't find much more about connection_string.
> 
> Hence my proposal to go on with #17912. It is old but the situation doesn't seem to have changed since then and the fix is not straightforward. At least not enough for me.

This seems to work to fix the issue on 9.3.1

**Comment by LukasJerabek** [comment]
Hi, I have got similar problem in postgresql 14 + timescaledb 2.9.1 with grafana 9.3.1.

Every midnight I get **DatasourceError alerts** (some repeats every day, others show occasionally, but it always happens at midnight). When examining the data, it does not look like there would be anything missing. However I once caught it exactly at midnight when I tried to run alert query and I have got this error in grafana window:
`database is locked, dashboard query runner forbidden 403`
I have also always find these in grafana.log:
```
logger=datasources t=2023-02-01T00:00:11.036218988+01:00 level=error msg="Failed getting data source" err="database is locked" uid=7uylVL07k id=0 name= orgId=1
logger=ngalert.scheduler rule_uid=A7JmCyOVk org_id=1 version=17 attempt=0 now=2023-02-01T00:00:00+01:00 t=2023-02-01T00:00:11.759236567+01:00 level=error msg="Failed to build rule evaluator" error="failed to build query 'A': database is locked"
logger=datasources t=2023-02-01T00:00:38.021860366+01:00 level=error msg="Failed getting data source" err="database is locked" uid=7uylVL07k id=0 name= orgId=1
logger=ngalert.scheduler rule_uid=0dM_9yOVz org_id=1 version=21 attempt=0 now=2023-02-01T00:00:00+01:00 t=2023-02-01T00:00:38.022852586+01:00 level=error msg="Failed to build rule evaluator" error="failed to build query 'A': database is locked"
logger=datasources t=2023-02-01T00:00:38.025806498+01:00 level=error msg="Failed getting data source" err="database is locked" uid=7uylVL07k id=0 name= orgId=1
logger=ngalert.scheduler rule_uid=FZizjydVz org_id=1 version=18 attempt=0 now=2023-02-01T00:00:00+01:00 t=2023-02-01T00:00:38.025876312+01:00 level=error msg="Failed to build rule evaluator" error="failed to build query 'A': database is locked"
logger=datasources t=2023-02-01T00:00:38.056431524+01:00 level=error msg="Failed getting data source" err="database is locked" uid=7uylVL07k id=0 name= orgId=1
logger=ngalert.scheduler rule_uid=mg-X2kK4k org_id=1 version=52 attempt=0 now=2023-02-01T00:00:00+01:00 t=2023-02-01T00:00:38.056540843+01:00 level=error msg="Failed to build rule evaluator" error="failed to build query 'A': database is locked"
logger=datasources t=2023-02-01T00:00:38.416748082+01:00 level=error msg="Failed getting data source" err="database is locked" uid=SEWMwHq7z id=0 name= orgId=1
logger=ngalert.scheduler rule_uid=m6gg1zKVz org_id=1 version=9 attempt=0 now=2023-02-01T00:00:00+01:00 t=2023-02-01T00:00:38.416899188+01:00 level=error msg="Failed to build rule evaluator" error="failed to build query 'A': database is locked"
logger=ngalert.multiorg.alertmanager t=2023-02-01T00:00:41.828170434+01:00 level=error msg="error while synchronizing Alertmanager orgs" error="database is locked"
logger=ngalert.scheduler rule_uid=gl9Jqwd4z org_id=1 version=63 attempt=0 now=2023-02-01T00:00:00+01:00 t=2023-02-01T00:00:50.682519118+01:00 level=error msg="Failed to evaluate rule" error="input data must be a wide series but got type long (input refid)" duration=0s
logger=ngalert.sender.router rule_uid=m6gg1zKVz org_id=1 t=2023-02-01T00:00:51.489028844+01:00 level=info msg="Sending alerts to local notifier" count=1
logger=ngalert.state.historian rule_uid=0dM_9yOVz org_id=1 t=2023-02-01T00:00:51.876814852+01:00 level=error msg="Error saving alert annotation batch" error="database is locked"
```

I was not sure if the issue is on timescaledb side or grafana, but I have noticed grafana has some troubles with this already. Could this be it is not just sqllite problem? I wonder why is it alway at midnight, does grafana do anything particular at midnight that may cause that?

**Comment by LukasJerabek** [comment]
> Hi, I have got similar problem in postgresql 14 + timescaledb 2.9.1 with grafana 9.3.1.
> 
> Every midnight I get **DatasourceError alerts** (some repeats every day, others show occasionally, but it always happens at midnight). When examining the data, it does not look like there would be anything missing. However I once caught it exactly at midnight when I tried to run alert query and I have got this error in grafana window: `database is locked, dashboard query runner forbidden 403` I have also always find these in grafana.log:
> 
> ```
> logger=datasources t=2023-02-01T00:00:11.036218988+01:00 level=error msg="Failed getting data source" err="database is locked" uid=7uylVL07k id=0 name= orgId=1
> logger=ngalert.scheduler rule_uid=A7JmCyOVk org_id=1 version=17 attempt=0 now=2023-02-01T00:00:00+01:00 t=2023-02-01T00:00:11.759236567+01:00 level=error msg="Failed to build rule evaluator" error="failed to build query 'A': database is locked"
> logger=datasources t=2023-02-01T00:00:38.021860366+01:00 level=error msg="Failed getting data source" err="database is locked" uid=7uylVL07k id=0 name= orgId=1
> logger=ngalert.scheduler rule_uid=0dM_9yOVz org_id=1 version=21 attempt=0 now=2023-02-01T00:00:00+01:00 t=2023-02-01T00:00:38.022852586+01:00 level=error msg="Failed to build rule evaluator" error="failed to build query 'A': database is locked"
> logger=datasources t=2023-02-01T00:00:38.025806498+01:00 level=error msg="Failed getting data source" err="database is locked" uid=7uylVL07k id=0 name= orgId=1
> logger=ngalert.scheduler rule_uid=FZizjydVz org_id=1 version=18 attempt=0 now=2023-02-01T00:00:00+01:00 t=2023-02-01T00:00:38.025876312+01:00 level=error msg="Failed to build rule evaluator" error="failed to build query 'A': database is locked"
> logger=datasources t=2023-02-01T00:00:38.056431524+01:00 level=error msg="Failed getting data source" err="database is locked" uid=7uylVL07k id=0 name= orgId=1
> logger=ngalert.scheduler rule_uid=mg-X2kK4k org_id=1 version=52 attempt=0 now=2023-02-01T00:00:00+01:00 t=2023-02-01T00:00:38.056540843+01:00 level=error msg="Failed to build rule evaluator" error="failed to build query 'A': database is locked"
> logger=datasources t=2023-02-01T00:00:38.416748082+01:00 level=error msg="Failed getting data source" err="database is locked" uid=SEWMwHq7z id=0 name= orgId=1
> logger=ngalert.scheduler rule_uid=m6gg1zKVz org_id=1 version=9 attempt=0 now=2023-02-01T00:00:00+01:00 t=2023-02-01T00:00:38.416899188+01:00 level=error msg="Failed to build rule evaluator" error="failed to build query 'A': database is locked"
> logger=ngalert.multiorg.alertmanager t=2023-02-01T00:00:41.828170434+01:00 level=error msg="error while synchronizing Alertmanager orgs" error="database is locked"
> logger=ngalert.scheduler rule_uid=gl9Jqwd4z org_id=1 version=63 attempt=0 now=2023-02-01T00:00:00+01:00 t=2023-02-01T00:00:50.682519118+01:00 level=error msg="Failed to evaluate rule" error="input data must be a wide series but got type long (input refid)" duration=0s
> logger=ngalert.sender.router rule_uid=m6gg1zKVz org_id=1 t=2023-02-01T00:00:51.489028844+01:00 level=info msg="Sending alerts to local notifier" count=1
> logger=ngalert.state.historian rule_uid=0dM_9yOVz org_id=1 t=2023-02-01T00:00:51.876814852+01:00 level=error msg="Error saving alert annotation batch" error="database is locked"
> ```
> 
> I was not sure if the issue is on timescaledb side or grafana, but I have noticed grafana has some troubles with this already.

also solved by 
`sudo sqlite3 /var/lib/grafana/grafana.db 'pragma journal_mode=wal;'`
Having ability to enable wal via config is a great idea. Thank you!

**Comment by slappey-ibkr** [comment]
Today we found majority of our dashboards somehow anonymously reverted to a previous version.  Under version history, the Updated by field is blank and the Notes field is blank.  It does not state "Restored from version xx".  I had to manually go to every dashboard and restore from previous version.  The only thing I could find in the logs were a few vague "database is locked" messages.  What process could do this?  I need to ensure this doesn't happen again.

**Comment by MiddleMan5** [comment]
We're also seeing a similar problem. Setting the `wal` configuration in the grafana settings does not seem to fix this issue

Environment:

- Grafana v10.1.1 with ngalert and oncall plugin
- sqlite3 with wal enabled
- Periodic alerts that database is locked
- Periodically unable to save dashboards due to error:
`logger=context userId=2 orgId=1 uname=quinn.mikelson@domain.com t=2023-10-04T17:56:13.449217515Z level=error msg="Failed to save dashboard" error="saving dashboard failed: database is locked" remote_addr=123.45.67.89 traceID=
`

![image](https://github.com/grafana/grafana/assets/12738069/6043fc71-e4d9-48c3-bbb0-81ec51ddb8a1)

**Comment by uluzox** [comment]
I had the same issue and found that my storage provider Azure File was not working with the Helm Chart. Switching to Azure Disk storage resolved the issue

**Comment by Oneoldmanz** [comment]
It's still relevant.
sudo sqlite3 /var/lib/grafana/grafana.db 'pragma journal_mode=wal' does not fix the situation.
Grafana v10.1.4 (a676a96d91)
endpoint telegram

**Comment by YourSandwich** [comment]
Same issue for me with the latest release

**Comment by shaunlowis** [comment]
At the time of writing this is still unresolved. I've got a samba server containing the `grafana.db`. I've mounted this and set up a dockerfile and docker-compose.yml. When I run this locally, it works fine and loads the `grafana.db` perfectly with docker-compose and also uses the `grafana.ini` file as expected. When I run this as a service in a docker swarm, I get the database locking issue.

A solution seems to be to migrate to postgreSQL, for which this tool exists: https://github.com/wbh1/grafana-sqlite-to-postgres
but isn't being maintained. Will attempt and report back.

**Comment by DanCech** [comment]
While WAL mode is helpful, it seems like the _busy_timeout parameter may also help alleviate the issues. It sets the sqlite [busy timeout](https://www.sqlite.org/c3ref/busy_timeout.html) to the specified value, which will cause sqlite to retry transactions before giving up and returning the busy aka "database is locked" error.

You can set that and WAL mode (note that once you switch a given sqlite db to WAL mode you cannot switch back) via the following database config:
```
[database]
url = sqlite3://localhost?_busy_timeout=500
wal = true
```
Other sqlite-specific settings  like `path` will continue to work as normal if you have your sqlite db in a non-standard location.

**Comment by goncalo-oliveira** [comment]
I was having a similar situation while running Grafana with SQLite over an Azure File Share. The WAL mode did not resolve the problem.

Setting `nobrl` on the [mount options](https://learn.microsoft.com/en-us/troubleshoot/azure/azure-kubernetes/storage/mountoptions-settings-azure-files) sorted it out for me.

> This setting prevents sending byte range lock requests to the server. It's necessary for certain applications that break with cifs-style mandatory byte range locks. Most cifs servers don't yet support requesting advisory byte range locks. If an application doesn't use this setting and breaks with cifs-style mandatory byte range locks, error messages such as Error: SQLITE_BUSY: database is locked might occur.

**Comment by kong62** [comment]
Version 10.3.3
```
Aug 20 03:58:09 wireguard-company grafana[4833]: logger=ngalert.sender.router rule_uid=effa3239-965a-4149-aaa1-1bed1b293801 org_id=1 t=2024-08-
20T03:58:09.811681316+08:00 level=info msg="Sending alerts to local notifier" count=4
Aug 20 03:59:03 wireguard-company grafana[4833]: logger=ngalert.sender.router rule_uid=effa3239-965a-4149-aaa1-1bed1b293801 org_id=1 t=2024-08-
20T03:59:03.077098246+08:00 level=info msg="Sending alerts to local notifier" count=4
Aug 20 04:00:03 wireguard-company grafana[4833]: logger=ngalert.sender.router rule_uid=effa3239-965a-4149-aaa1-1bed1b293801 org_id=1 t=2024-08-
20T04:00:03.281074601+08:00 level=info msg="Sending alerts to local notifier" count=4
Aug 20 04:01:06 wireguard-company grafana[4833]: logger=ngalert.sender.router rule_uid=effa3239-965a-4149-aaa1-1bed1b293801 org_id=1 t=2024-08-
20T04:01:06.110715113+08:00 level=info msg="Sending alerts to local notifier" count=4
Aug 20 04:01:44 wireguard-company grafana[4833]: logger=secrets.kvstore t=2024-08-20T04:01:44.154890932+08:00 level=error msg="error getting se
cret value" orgId=1 type=datasource namespace=prometheus err="database is locked"
Aug 20 04:01:44 wireguard-company grafana[4833]: logger=sqlstore.transactions rule_uid=bfe379ec-019b-4103-8766-179103749620 org_id=1 t=2024-08-
20T04:01:44.154988772+08:00 level=info msg="Database locked, sleeping then retrying" error="database is locked" retry=1 code="database is locke
d"
Aug 20 04:01:47 wireguard-company grafana[4833]: logger=ngalert.sender.router rule_uid=bfe379ec-019b-4103-8766-179103749620 org_id=1 t=2024-08-
20T04:01:47.214549044+08:00 level=info msg="Sending alerts to local notifier" count=1
Aug 20 04:02:02 wireguard-company grafana[4833]: logger=ngalert.sender.router rule_uid=effa3239-965a-4149-aaa1-1bed1b293801 org_id=1 t=2024-08-
20T04:02:02.106930908+08:00 level=info msg="Sending alerts to local notifier" count=4
Aug 20 04:02:11 wireguard-company grafana[4833]: logger=datasources t=2024-08-20T04:02:11.153058964+08:00 level=error msg="Failed getting data 
source" err="database is locked" uid=a3631df0-bcfa-494b-b0d2-36becf816f39 id=0 name= orgId=1
Aug 20 04:02:11 wireguard-company grafana[4833]: logger=sqlstore.transactions rule_uid=b0e91da2-fe24-4654-98a1-acf7266f2f3a org_id=1 t=2024-08-
20T04:02:11.153150476+08:00 level=info msg="Database locked, sleeping then retrying" error="database is locked" retry=1 code="database is locke
d"
Aug 20 04:02:11 wireguard-company grafana[4833]: logger=ngalert.scheduler rule_uid=b0e91da2-fe24-4654-98a1-acf7266f2f3a org_id=1 version=37 fin
gerprint=0d7fb7c389c8aa05 attempt=1 now=2024-08-20T04:02:00+08:00 rule_uid=b0e91da2-fe24-4654-98a1-acf7266f2f3a org_id=1 t=2024-08-20T04:02:11.
153205628+08:00 level=error msg="Failed to build rule evaluator" error="failed to build query 'A': [sqlstore.max-retries-reached] retry 1: data
base is locked"
``` q

**Comment by xurengithub** [comment]
My Grafana service instance crashed directly
Version:9.4.3

error log

``` logger=datasources t=2024-11-19T23:04:10.58143652+08:00 level=error msg="Failed getting data source" err="database is locked" uid=9vD9V7BIz id=0 name= orgId=1
logger=sqlstore.transactions rule_uid=hJLDC9CSk org_id=1 t=2024-11-19T23:04:10.58149441+08:00 level=info msg="Database locked, sleeping then retrying" error="database is locked" retry=1 code="database is locked"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2e13658]


goroutine 13398698 [running]:
github.com/grafana/grafana/pkg/services/datasources/service.(*CacheServiceImpl).GetDatasourceByUID(0xc001a40ab0, {0x4e57f08, 0xc001f3a270}, {0xc002a6e5e0, 0x9}, 0xc001b51880, 0x0)
        /drone/src/pkg/services/datasources/service/cache.go:104 +0x3d8
github.com/grafana/grafana/pkg/services/ngalert/eval.getExprRequest({{0x4e57f08?, 0xc001f3a270?}, 0xc001b51880?}, {0xc00359aa80, 0x3, 0xe0?}, {0x4e3b098, 0xc001a40ab0})
        /drone/src/pkg/services/ngalert/eval/eval.go:273 +0x403
github.com/grafana/grafana/pkg/services/ngalert/eval.(*evaluatorImpl).Create(0xc00168da40, {{0x4e57f08?, 0xc001f3a270?}, 0xc001b51880?}, {{0xc002a6e3d0, 0x1}, {0xc00359aa80, 0x3, 0x4}})
        /drone/src/pkg/services/ngalert/eval/eval.go:630 +0x86
github.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func3({0x4e57f08, 0xc001f3a270}, 0x4?, 0xc002b45c88, {0x4e683d0, 0xc002b30dd0})
        /drone/src/pkg/services/ngalert/schedule/schedule.go:362 +0x4b5
github.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func5.2(0x48?)
        /drone/src/pkg/services/ngalert/schedule/schedule.go:499 +0x951
github.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func4(0xc00eb0c640)
        /drone/src/pkg/services/ngalert/schedule/schedule.go:428 +0x5b
github.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine.func5(0xc003d51d46?, 0xc001287450, {0x1?, {0xc0019d38f0?, 0x9?}}, 0xc002b45c88, 0xc003697e88, 0xc0019d3d30, {0x4e6ad90, 0xc002f18090},
 ...)
        /drone/src/pkg/services/ngalert/schedule/schedule.go:472 +0x1d6
github.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).ruleRoutine(0xc001287450, {0x4e59750, 0xc002267f20}, {0xc0000ac280?, {0xc0019d38f0?, 0xc0031e29c0?}}, 0xc0031e2a20, 0xc0031e2a80)
        /drone/src/pkg/services/ngalert/schedule/schedule.go:505 +0x825
github.com/grafana/grafana/pkg/services/ngalert/schedule.(*schedule).processTick.func1()
        /drone/src/pkg/services/ngalert/schedule/schedule.go:247 +0x3c
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /opt/drone/gomodcache/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
        /opt/drone/gomodcache/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:72 +0xa5
``` q

**Comment by Gitkingly** [comment]
Unbelievable that this problem still exists today  when I installed with helm-charts-prometheus-stack-70.7.0. It used to work but there were errors some day. This unpredictable behavior is unacceptable:

logger=grafana-apiserver t=2025-06-11T06:02:42.040221233Z level=info msg="Adding GroupVersion iam.grafana.app v0alpha1 to ResourceManager"
logger=grafana-apiserver t=2025-06-11T06:02:42.042923796Z level=info msg="Adding GroupVersion notifications.alerting.grafana.app v0alpha1 to ResourceManager"
logger=grafana-apiserver t=2025-06-11T06:02:42.04380712Z level=info msg="Adding GroupVersion userstorage.grafana.app v0alpha1 to ResourceManager"
logger=grafana-apiserver t=2025-06-11T06:02:42.044592035Z level=info msg="Adding GroupVersion playlist.grafana.app v0alpha1 to ResourceManager"
logger=grafana-apiserver t=2025-06-11T06:02:42.098648144Z level=info msg="StorageObjectCountTracker pruner is exiting"
logger=sqlstore.transactions t=2025-06-11T06:02:42.272843425Z level=info msg="Database locked, sleeping then retrying" error="database is locked" retry=0 code="database is locked"
logger=sqlstore.transactions t=2025-06-11T06:02:42.284356345Z level=info msg="Database locked, sleeping then retrying" error="database is locked" retry=1 code="database is locked"
logger=sqlstore.transactions t=2025-06-11T06:02:42.295614143Z level=info msg="Database locked, sleeping then retrying" error="database is locked" retry=2 code="database is locked"
logger=sqlstore.transactions t=2025-06-11T06:02:42.307019429Z level=info msg="Database locked, sleeping then retrying" error="database is locked" retry=3 code="database is locked"
logger=sqlstore.transactions t=2025-06-11T06:02:42.318254203Z level=info msg="Database locked, sleeping then retrying" error="database is locked" retry=4 code="database is locked"
logger=plugin.angulardetectorsprovider.dynamic t=2025-06-11T06:02:42.319405645Z level=error msg="Error while updating detectors" error="set last updated: database is locked"
logger=plugin.angulardetectorsprovider.dynamic t=2025-06-11T06:02:42.319432686Z level=info msg="Patterns update finished" duration=672.520476ms
logger=plugin.backgroundinstaller t=2025-06-11T06:02:42.365886478Z level=info msg="Installing plugin" pluginId=grafana-lokiexplore-app version=
logger=sqlstore.transactions t=2025-06-11T06:02:46.668348783Z level=info msg="Database locked, sleeping then retrying" error="database is locked" retry=1 code="database is locked"
logger=auth t=2025-06-11T06:02:46.668489798Z level=error msg="Failed to lock and execute cleanup of expired auth token" error="[sqlstore.max-retries-reached] retry 1: database is locked"
logger=sqlstore.transactions t=2025-06-11T06:02:46.668794151Z level=info msg="Database locked, sleeping then retrying" error="database is locked" retry=1 code="database is locked"
logger=serviceaccounts t=2025-06-11T06:02:46.668847769Z level=error msg="Failed to lock and execute the migration of API keys to service accounts" error="[sqlstore.max-retries-reached] retry 1: database is locked"
logger=secret.migration t=2025-06-11T06:02:46.669497939Z level=error msg="Server lock for secret migration already exists"

**Comment by Rohlik** [comment]
For us, it was solved by upgrading to Grafana `13.0.1`.
